### PR TITLE
Support materialized CTEs

### PR DESCRIPTION
**Pull request by maartenvanvliet**
companion PR to https://github.com/elixir-ecto/ecto/pull/4066

### CODE REVIEW

File: lib/ecto/adapters/postgres/connection.ex
Comment (warmwaffles) [inline lib/ecto/adapters/postgres/connection.ex:415]: If this gets accepted, I'll add support for this to `ecto_sqlite3` it supports materialized CTEs https://www.sqlite.org/lang_with.html

File: lib/ecto/adapters/myxql/connection.ex
Comment (v0idpwn) [inline lib/ecto/adapters/myxql/connection.ex:317]:
```suggestion
      error!(query, "MySQL adapter does not support materialized CTEs")
```

File: lib/ecto/adapters/tds/connection.ex
Comment (v0idpwn) [inline lib/ecto/adapters/tds/connection.ex:440]:
```suggestion
    defp cte_expr({_name, true, _cte}, _sources, _query) do
      error!(query, "Tds adapter does not support materialized CTEs")
    end

    defp cte_expr({name, _materialize, cte}, sources, query) do
```

File: lib/ecto/adapters/postgres/connection.ex
Comment (greg-rychlewski) [inline lib/ecto/adapters/postgres/connection.ex:414]:
```suggestion
    defp cte_expr({name, materialized, cte}, sources, query) do
      materialized_opt =
        case materialized do
          nil -> ""
          true -> "MATERIALIZED"
          false -> "NOT MATERIALIZED"  
        end
        
        [quote_name(name), " AS ", materialized_opt, cte_query(cte, sources, query)]
```

Based on what we talked about in the ecto PR. also the variable is `materliazed` instead of `materialize` so better to be consistent

File: lib/ecto/adapters/myxql/connection.ex
Comment (greg-rychlewski) [inline lib/ecto/adapters/myxql/connection.ex:316]:
```suggestion
    defp cte_expr({_name, _materialized, _cte}, _sources, _query) do when is_boolean(materialized)
```

File: lib/ecto/adapters/myxql/connection.ex
Comment (greg-rychlewski) [inline lib/ecto/adapters/myxql/connection.ex:320]:
```suggestion
    defp cte_expr({name, nil, cte}, sources, query) do
```

File: lib/ecto/adapters/tds/connection.ex
Comment (greg-rychlewski) [inline lib/ecto/adapters/tds/connection.ex:441]: same comments as for mysql

File: test/ecto/adapters/postgres_test.exs
Comment (greg-rychlewski) [inline test/ecto/adapters/postgres_test.exs:158]:
```suggestion

```

File: test/ecto/adapters/postgres_test.exs
Comment (greg-rychlewski) [inline test/ecto/adapters/postgres_test.exs:160]:
```suggestion

```

File: test/ecto/adapters/postgres_test.exs
Comment (greg-rychlewski) [inline test/ecto/adapters/postgres_test.exs:159]: we should also add a test for materialized = false.

**Comment by greg-rychlewski** [comment]
if you're able to point this at your Ecto PR, that will help us verify the tests are working :).

**Review by greg-rychlewski — APPROVED** [review]

**Comment by greg-rychlewski** [comment]
Looks good to me thank you. I'll just wait for one more approval from @v0idpwn or @josevalim

**Review by v0idpwn — APPROVED** [review]

**Pull request merged**
